include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,622 @@
+                              string_view request_body, const Pairs &request_trailers,
+                              int timeout_milliseconds, HttpCallToken *token_ptr) = 0;
+};
+
+struct GrpcCallInterface {
+  /**
+   * Make a gRPC call.
+   * @param grpc_service is proxy-specific metadata describing the service (e.g. security certs).
+   * @param service_name the name of the gRPC service.
+   * @param method_name the gRPC method name.
+   * @param request the serialized request.
+   * @param initial_metadata the initial metadata.
+   * @param timeout a timeout in milliseconds.
+   * @param token_ptr contains a pointer to a location to store the token which will be used with
+   * the corresponding onGrpc and grpc calls.
+   */
+  virtual WasmResult grpcCall(string_view /* grpc_service */, string_view /* service_name */,
+                              string_view /* method_name */, string_view /* request */,
+                              const Pairs & /* initial_metadata */,
+                              std::chrono::milliseconds & /* timeout */,
+                              GrpcToken * /* token_ptr */) = 0;
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcStream.
+   * For call the same as grpcCancel. For Stream, close the stream: outstanding peer messages may
+   * still arrive.
+   */
+  virtual WasmResult grpcClose(GrpcToken /* token */) = 0;
+
+  /**
+   * Cancel a gRPC stream or call.  No more calls will occur.
+   * @param token is a token returned from grpcSream or grpcCall.
+   * For Call, cancel on call. For Strea, reset the stream: no further callbacks will arrive.
+   */
+  virtual WasmResult grpcCancel(GrpcToken /* token */) = 0;
+};
+
+struct GrpcStreamInterface {
+  /**
+   * Open a gRPC stream.
+   * @param grpc_service is proxy-specific metadata describing the service (e.g. security certs).
+   * @param service_name the name of the gRPC service.
+   * @param method_name the gRPC method name.
+   * @param token_ptr contains a pointer to a location to store the token which will be used with
+   * the corresponding onGrpc and grpc calls.
+   */
+  virtual WasmResult grpcStream(string_view grpc_service, string_view service_name,
+                                string_view method_name, GrpcToken *token_ptr) = 0;
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcSream.
+   * @param message is a (serialized) message to be sent.
+   * @param end_stream indicates that the stream is now end_of_stream (e.g. WriteLast() or
+   * WritesDone).
+   */
+  virtual WasmResult grpcSend(GrpcToken token, string_view message, bool end_stream) = 0;
+
+  // See GrpcCallInterface.
+  virtual WasmResult grpcClose(GrpcToken token) = 0;
+
+  // See GrpcCallInterface.
+  virtual WasmResult grpcCancel(GrpcToken token) = 0;
+};
+
+struct MetricsInterface {
```

Comment:
Can you mention here that the spec will define the contract around what operations are supported for metric types, etc. and to consult that doc?

Reply:
done.